### ISSUE TITLE
[$500] [Distance] - Map does not zoom out to show the entire route when the map window is reduced

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#26724]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
**Issue found when executing PR** https://github.com/Expensify/App/pull/26591

## Action Performed:

1. Launch Expensify.
2. Open the global create menu and select Request money.
3. Select the Distance tab if it is not already selected.
4. Add the Start (1 Telegraph Hill Blvd) and Finish (88 Kearny St. San Francisco) waypoints.
5. Resize the browser window vertically by either reducing or increasing the height of the window.

## Expected Result:
The map zooms out to show the entire route when the map window is reduced

## Actual Result:
The map does not zoom out to show the entire route when the map window is reduced.
On Android app, the map zooms out automatically

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome
- [ ] MacOS / Desktop

**Version Number:** 1.3.63-0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/1c3129c0-9a22-40cd-b9e0-2d2c32810e35

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01d3c0d2628f023ef1</li>
        <li>Upwork Job ID: 1701023325165142016</li>
        <li>Last Price Increase: 2023-10-21</li>
    </ul>
</details>

**Comment by jliexpensify** [comment]
So the issue here is that on New.Dot (non-Android), you add a 3rd stop and the green line disappears and the map zooms out, almost in preparation for you to add that 3rd stop:

![image](https://github.com/Expensify/App/assets/30815269/992d0dee-f3ba-40e2-a51d-9d18cffca4a7)

Whereas in Android it doesn't - the line remains and doesn't really zoom.

**Comment by jliexpensify** [comment]
Ok, I updated my Android app and can confirm that this is an app only issue - Chrome is fine, and follows New.Dot (web) behaviours.

**Comment by jliexpensify** [comment]
@lanitochka17 can you clarify what you expect the behaviour to be? Should we be seeing a green line or not?

**Comment by jliexpensify** [comment]
Here's another regression (?) @hayata-suenaga and @situchan - it looks like the Android app shows a green line when adding a 3rd location, whereas Android Chrome and Web do not.

**Comment by jliexpensify** [comment]
@hayata-suenaga it doesn't look like a job was created when the `External` label was added so re-added. Also, I wanted to confirm that this isn't an issue on Android? From my initial testing, [it seems like it could be an issue on the Android app too](https://github.com/Expensify/App/issues/26724#issuecomment-1705917578).

**Comment by hayata-suenaga** [comment]
Thank you for reapplying the label 🙇 And thank you for testing on Android

**Comment by akinwale** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Map does not zoom out to show entire route when the map window is reduced, and the route re-renders when the map is resized.

### What is the root cause of that problem?
This is actually being caused by the route being cleared whenever a stop is added to the transaction. This creates a brief moment where the direction coordinate passed to the map is null, which causes it to clear, and then re-render when the route is eventually recalculated.

### What changes do you think we should make in order to solve the problem?
**Solution 1**
Do not clear the route when a stop is added to the transaction. Instead check if there are at least 2 validated waypoints before clearing the route.

**Solution 2**
If there at least 2 validated waypoints in DistanceRequest but directionCoordinates is null, do not update the prop. Otherwise only update the directionCoordinates prop if there is only one waypoint and the route calculation is null.

### What alternative solutions did you explore? (Optional)
None.

**Comment by situchan** [comment]
@akinwale thanks for the proposal.

>  the map does not automatically readjust the bounds on web as it does on the native iOS and Android platforms

Can we try to fix this in upstream (mapbox web sdk)?

**Comment by akinwale** [comment]
It looks like this is something that we're meant to implement ourselves when using Mapbox on the web, not exactly an issue. For the web, [Mapbox GL JS](https://github.com/mapbox/mapbox-gl-js/) is the underlying Mapbox implementation used.

Here are some of the issues related to what we're trying to achieve.
https://github.com/mapbox/mapbox-gl-js/issues/2090 - from 2016.
https://github.com/mapbox/mapbox-gl-js/issues/8920 - from 2019.
https://github.com/mapbox/mapbox-gl-js/issues/11004 - the most recent I could find from 2021.

The consensus seems that if the map container is resized, you have to call the `map.resize()` method first before performing any new operations on the map.

I use a boolean state variable to track whether or not to call the `fitBounds` method since the state update is done afterwards. Trying to call `fitBounds` in the map resize event handler (`mapRef.on('resize', ...`) does not work as expected since it seems the map resize operation is not complete at that point.

**Comment by situchan** [comment]
As this won't be most likely fixed in upstream, need to test proposed solution of local fix.

**Comment by akinwale** [comment]
> As this won't be most likely fixed in upstream, need to test proposed solution of local fix.

The route refresh issue is fixed in https://github.com/Expensify/App/pull/27496. So only the bounds functionality needs to be tested here.

Also, it looks like map resize will be added to https://github.com/Expensify/App/issues/26540 so it makes sense to wait for the [PR](https://github.com/Expensify/App/pull/27574/files) on that to be merged before we proceed here because it makes use of a ResizeObserver to trigger `map.resize`.

**Comment by akinwale** [comment]
@situchan https://github.com/Expensify/App/pull/27574 has been merged and deployed to staging, so I think we can move forward here.

**Comment by situchan** [comment]
I think we can wait a bit more to hit production. I hope that won't happen but revert case due to regression.

**Comment by hayata-suenaga** [comment]
Interesting that we have to call `resize` every time the map size changes...

by the way, why are we waiting on [this PR](https://github.com/Expensify/App/pull/27574)?

**Comment by akinwale** [comment]
> Interesting that we have to call `resize` every time the map size changes...
> 
> by the way, why are we waiting on [this PR](https://github.com/Expensify/App/pull/27574)?

It implements calling `map.resize()` using the `ResizeObserver` which is also done in my [proposal](https://github.com/Expensify/App/issues/26724#issuecomment-1713001807). I recommended waiting for it to be merged to avoid a duplication of effort. I feel like we can move forward at this point, but we can also wait for the regression tests to be complete according to @situchan's recommendation.

**Comment by hayata-suenaga** [comment]
yes let's wait for the PR to be deployed to production

**Comment by akinwale** [comment]
@situchan https://github.com/Expensify/App/pull/27574 has been deployed to production.

**Comment by jliexpensify** [comment]
I'm currently on a bad internet connection, but I tested this (v 73-0 Chrome) and think this is now fixed? I'm seeing the map points properly move now cc @hayata-suenaga @situchan

**Comment by akinwale** [comment]
@jliexpensify The steps to reproduce have changed since the Add Stop behaviour has been modified. It can be reproduced if the browser window is vertically resized, of if the map is resized due to a change in responsive layout.

**Comment by hayata-suenaga** [comment]
@akinwale 

The issue seems like to be still present

https://github.com/Expensify/App/assets/98560306/c744c9d3-29e2-41e6-b273-be897a1e353c

**Comment by akinwale** [comment]
> @akinwale
> 
> The issue seems like to be still present

Looks like it. The changes in my proposal should still work in this case. We can move forward when @situchan is ready.

**Comment by hayata-suenaga** [comment]
ah I somehow thought the PR is already deployed to production 😓

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by akinwale** [comment]
> Issue not reproducible during KI retests. (First week)

The reproduction steps have changed.

**Comment by jliexpensify** [comment]
Hmm so what's happening here? Are we waiting on @situchan to do something? [Going off this comment.](https://github.com/Expensify/App/issues/26724#issuecomment-1734814174)

**Comment by situchan** [comment]
@akinwale can you update your proposal based on latest codebase? So remove deprecated solutions and update demo video to reflect browser resizing instead of adding new route.

**Comment by akinwale** [comment]
> @akinwale can you update your proposal based on latest codebase? So remove deprecated solutions and update demo video to reflect browser resizing instead of adding new route.

@situchan Updated the proposal and demo video.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by hayata-suenaga** [comment]
>> Issue not reproducible during KI retests. (First week)

> The reproduction steps have changed.

@mvtglobally Did you follow the new reproduction steps?

**Comment by mvtglobally** [comment]
@hayata-suenaga are steps updated in the original top report? Or it's in the comments? 
We usually follow the original report steps only.

**Comment by hayata-suenaga** [comment]
@akinwale or @situchan can you find the latest test steps? 🙇

**Comment by akinwale** [comment]
> @akinwale or @situchan can you find the latest test steps? 🙇

1. Launch Expensify.
2. Open the global create menu and select Request money.
3. Select the Distance tab if it is not already selected.
4. Add the Start (1 Telegraph Hill Blvd) and Finish (88 Kearny St. San Francisco) waypoints.
5. Resize the browser window vertically by either reducing or increasing the height of the window.

**Comment by hayata-suenaga** [comment]
thank you so much @akinwale!

I updated the original top report with the test steps above. @mvtglobally please retest again 🙇

**Comment by kbecciv** [comment]
Issue is not reproduced with provided steps here https://github.com/Expensify/App/issues/26724#issuecomment-1748145650

https://github.com/Expensify/App/assets/93399543/1ac0afcf-0318-4e44-8fa8-b7cf5a592d54

**Comment by akinwale** [comment]
@hayata-suenaga From the test video above, it looks like the map now has a fixed height which doesn't change. Is this intentional? If so, then this issue is no longer relevant.

**Comment by situchan** [comment]
> @hayata-suenaga From the test video above, it looks like the map now has a fixed height which doesn't change. Is this intentional? If so, then this issue is no longer relevant.

No, it will be fixed in https://github.com/Expensify/App/issues/28911.
I think we need to put this on hold until https://github.com/Expensify/App/pull/28922 is merged.

**Comment by jliexpensify** [comment]
Changing to Weekly as it's on HOLD

**Comment by neil-marcellini** [comment]
I'm taking this over for Hayata as decided here (internally) in Slack. Sounds like we're holding for [Fix map preview not extended to full view](https://github.com/Expensify/App/pull/28922#top) right now which makes sense.

**Comment by neil-marcellini** [comment]
I updated this issue to be held on the issue that PR fixes ^ since we usually hold on the issue rather than the PR.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Third week)

**Comment by neil-marcellini** [comment]
So longer on HOLD. @jliexpensify would you please test if it's still reproducible?

**Comment by akinwale** [comment]
> So longer on HOLD. @jliexpensify would you please test if it's still reproducible?

Still reproducible on the latest main branch.

**Comment by jliexpensify** [comment]
Tested on Staging v1.3.88-3:

https://github.com/Expensify/App/assets/30815269/150c4fba-ed26-4254-ba6e-529c49354bf0

This is the same as Android (Chrome) and Android app - the map auto-zooms out.

@akinwale can you share a video?

**Comment by akinwale** [comment]
@jliexpensify The steps to reproduce have changed.

Context: https://github.com/Expensify/App/issues/26724#issuecomment-1732826149

Video: https://github.com/Expensify/App/issues/26724#issuecomment-1734676514

Updated steps: https://github.com/Expensify/App/issues/26724#issuecomment-1748145650

**Comment by jliexpensify** [comment]
Ah sorry @akinwale - here's a re-test with the new steps

https://github.com/Expensify/App/assets/30815269/78263363-96ef-4f59-bfb8-1b23485028af

Let me know if I've reproduced this incorrectly!

**Comment by akinwale** [comment]
@jliexpensify Looks like it's not as noticeable as it was previously. Can you try to manually zoom in on the map using your mouse wheel and then resize the browser window to see if it adjusts?

**Comment by jliexpensify** [comment]
Looks like no adjustment if zoomed in:

https://github.com/Expensify/App/assets/30815269/114ede95-fe09-4060-a08b-b7d8eab6497c

**Comment by akinwale** [comment]
> Looks like no adjustment if zoomed in:

Yeah, so we'll still need to fix that if it's considered worth fixing.

cc @neil-marcellini

**Comment by neil-marcellini** [comment]
I don't think it's a problem if it doesn't adjust while zoomed. I remember that at one point we considered making the map non-interactive on this page, but I'm not passionate about it. I don't see any problems now, so I think we can close this. Do you agree @jliexpensify?

**Comment by jliexpensify** [comment]
Yeah, I agree - I don't think it's a huge deal. Closing!

**Comment by situchan** [comment]
Should https://github.com/Expensify/App/issues/29469 be closed as well?

**Comment by jliexpensify** [comment]
Thanks @situchan , left a comment in that other GH!
